// xml-core/build/cjs/collection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Collection = void 0;
class Collection {
    constructor(items) {
        this.items = [];
        if (items) {
            this.items = items;
        }
    }
    get Count() {
        return this.items.length;
    }
    Item(index) {
        return this.items[index] || null;
    }
    Add(item) {
        this.items.push(item);
    }
    Pop() {
        return this.items.pop();
    }
    RemoveAt(index) {
        this.items = this.items.filter((item, index2) => index2 !== index);
    }
    Clear() {
        this.items = [];
    }
    GetIterator() {
        return this.items;
    }
    ForEach(cb) {
        this.GetIterator().forEach(cb);
    }
    Map(cb) {
        return new Collection(this.GetIterator().map(cb));
    }
    Filter(cb) {
        return new Collection(this.GetIterator().filter(cb));
    }
    Sort(cb) {
        return new Collection(this.GetIterator().sort(cb));
    }
    Every(cb) {
        return this.GetIterator().every(cb);
    }
    Some(cb) {
        return this.GetIterator().some(cb);
    }
    IsEmpty() {
        return this.Count === 0;
    }
}
exports.Collection = Collection;
