// xml-core/build/cjs/error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.XE = exports.XmlError = void 0;
function printf(text, ...args) {
    let msg = text.replace(/%([1-9]\d*)/g, (m, n) => {
        const idx = parseInt(n, 10) - 1;
        return args[idx] !== undefined ? args[idx] : '';
    });
    msg = msg.replace(/%%/g, '%');
    return msg;
}
function padNum(num, size) {
    let s = num + '';
    while (s.length < size) {
        s = '0' + s;
    }
    return s;
}
class XmlError {
    constructor(code, ...args) {
        this.prefix = 'XMLJS';
        this.code = code;
        this.name = this.constructor.name;
        const message = printf(xes[code], ...args);
        this.message = `${this.prefix}${padNum(code, 4)}: ${message}`;
        this.stack = new Error(this.message).stack;
    }
}
exports.XmlError = XmlError;
var XE;
(function (XE) {
    XE[XE["NONE"] = 0] = "NONE";
    XE[XE["NULL_REFERENCE"] = 1] = "NULL_REFERENCE";
    XE[XE["NULL_PARAM"] = 2] = "NULL_PARAM";
    XE[XE["DECORATOR_NULL_PARAM"] = 3] = "DECORATOR_NULL_PARAM";
    XE[XE["COLLECTION_LIMIT"] = 4] = "COLLECTION_LIMIT";
    XE[XE["METHOD_NOT_IMPLEMENTED"] = 5] = "METHOD_NOT_IMPLEMENTED";
    XE[XE["METHOD_NOT_SUPPORTED"] = 6] = "METHOD_NOT_SUPPORTED";
    XE[XE["PARAM_REQUIRED"] = 7] = "PARAM_REQUIRED";
    XE[XE["CONVERTER_UNSUPPORTED"] = 8] = "CONVERTER_UNSUPPORTED";
    XE[XE["ELEMENT_MALFORMED"] = 9] = "ELEMENT_MALFORMED";
    XE[XE["ELEMENT_MISSING"] = 10] = "ELEMENT_MISSING";
    XE[XE["ATTRIBUTE_MISSING"] = 11] = "ATTRIBUTE_MISSING";
    XE[XE["CONTENT_MISSING"] = 12] = "CONTENT_MISSING";
    XE[XE["CRYPTOGRAPHIC"] = 13] = "CRYPTOGRAPHIC";
    XE[XE["CRYPTOGRAPHIC_NO_MODULE"] = 14] = "CRYPTOGRAPHIC_NO_MODULE";
    XE[XE["CRYPTOGRAPHIC_UNKNOWN_TRANSFORM"] = 15] = "CRYPTOGRAPHIC_UNKNOWN_TRANSFORM";
    XE[XE["ALGORITHM_NOT_SUPPORTED"] = 16] = "ALGORITHM_NOT_SUPPORTED";
    XE[XE["ALGORITHM_WRONG_NAME"] = 17] = "ALGORITHM_WRONG_NAME";
    XE[XE["XML_EXCEPTION"] = 18] = "XML_EXCEPTION";
    XE[XE["KEY_INFO_CLAUSE_NOT_SUPPORTED"] = 19] = "KEY_INFO_CLAUSE_NOT_SUPPORTED";
})(XE || (exports.XE = XE = {}));
const xes = {};
xes[XE.NONE] = 'No description';
xes[XE.NULL_REFERENCE] = 'Null reference';
xes[XE.NULL_PARAM] = `'%1' has empty '%2' object`;
xes[XE.DECORATOR_NULL_PARAM] = `Decorator '%1' has empty '%2' parameter`;
xes[XE.COLLECTION_LIMIT] = `Collection of '%1' in element '%2' has wrong amount of items`;
xes[XE.METHOD_NOT_IMPLEMENTED] = 'Method is not implemented';
xes[XE.METHOD_NOT_SUPPORTED] = 'Method is not supported';
xes[XE.PARAM_REQUIRED] = `Required parameter is missing '%1'`;
xes[XE.CONVERTER_UNSUPPORTED] = 'Converter is not supported';
xes[XE.ELEMENT_MALFORMED] = `Malformed element '%1'`;
xes[XE.ELEMENT_MISSING] = `Element '%1' is missing in '%2'`;
xes[XE.ATTRIBUTE_MISSING] = `Attribute '%1' is missing in '%2'`;
xes[XE.CONTENT_MISSING] = `Content is missing in '%1'`;
xes[XE.CRYPTOGRAPHIC] = 'Cryptographic error: %1';
xes[XE.CRYPTOGRAPHIC_NO_MODULE] = 'WebCrypto module is not found';
xes[XE.CRYPTOGRAPHIC_UNKNOWN_TRANSFORM] = 'Unknown transform %1';
xes[XE.ALGORITHM_NOT_SUPPORTED] = `Algorithm is not supported '%1'`;
xes[XE.ALGORITHM_WRONG_NAME] = `Algorithm wrong name in use '%1'`;
xes[XE.XML_EXCEPTION] = 'XML exception: %1';
xes[XE.KEY_INFO_CLAUSE_NOT_SUPPORTED] = `KeyInfoClause type '%1' is not supported`;
